File: jorendorff+haskell/Pennies.hs
{-# LANGUAGE TypeFamilies #-}

import Minimax
import Vs

data Pennies = Pennies Int
  deriving Show

instance Game Pennies where
  type Move Pennies = Int
  start = Pennies 11
  moves (Pennies n) = filter (<= n) [1, 2, 3]
  applyMove (Pennies n) k = Pennies (n - k)
  scoreFinishedGame (Pennies 0) = 1

main = playHumanVsComputer bestMove (Pennies 14)

-- penniesSmartyPantsAI :: Pennies -> Move Pennies
-- penniesSmartyPantsAI =
--   bestMoveWithDepthLimit (\(Pennies n) -> if n `mod` 4 == 0 then 1.0 else -1.0) 2

File: jorendorff+rust/src/pennies.rs
// *** pennies.rs: the game of Pennies ****************************************

use minimax::Game;

#[derive(Debug, Clone, Copy)]
pub struct Pennies(i32);

impl Game for Pennies {
    type Move = i32;

    fn start() -> Pennies {
        Pennies(14)
    }
    fn moves(&self) -> Vec<i32> {
        let Pennies(n) = *self;
        (1..4).filter(|x| n - x >= 0).collect::<Vec<i32>>()
    }
    fn apply_move(&self, m: i32) -> Pennies {
        let Pennies(n) = *self;
        Pennies(n - m)
    }
    fn score_finished_game(&self) -> f64 {
        1.0
    }
}
